File: python/seren_agent/agent.py
"""Agent decorator for Seren templates.

The @agent decorator marks a function as a Seren agent template. Templates
must implement the contract: run(input: dict) -> dict

The decorator is backend-agnostic - the same code runs on Daytona, Modal,
E2B, or any other supported compute backend.
"""

from typing import Callable, Optional
from functools import wraps


def agent(
    name: str,
    description: str = "",
    price: str = "0.01",
    compute_backend: Optional[str] = None,
) -> Callable:
    """Decorator to mark a function as a Seren agent template.

    Args:
        name: Display name for the agent in the store
        description: What the agent does (shown in catalog)
        price: Price per invocation in USD (e.g., "0.05" for 5 cents)
        compute_backend: Preferred execution backend. If None, uses store default.
            Options: "daytona", "modal", "e2b", etc.

    Returns:
        Decorated function with _seren_agent metadata

    Raises:
        ValueError: If name is empty or price is invalid

    Example:
        @agent(name="Code Analyzer", price="0.10", description="Analyzes code quality")
        def run(input: dict) -> dict:
            code = input["code"]
            # ... analysis logic ...
            return {"score": 85, "issues": [...]}
    """
    # Validate required parameters
    if not name or not name.strip():
        raise ValueError("Agent name is required and cannot be empty")

    if not price or not price.strip():
        raise ValueError("Agent price is required and cannot be empty")

    # Validate price format (should be a valid decimal string)
    try:
        price_float = float(price)
        if price_float < 0:
            raise ValueError("Agent price cannot be negative")
    except ValueError as e:
        if "could not convert" in str(e).lower():
            raise ValueError(f"Agent price must be a valid number, got: {price}")
        raise

    def decorator(func: Callable) -> Callable:
        @wraps(func)
        def wrapper(input_data: dict) -> dict:
            return func(input_data)

        # Attach metadata for the Seren CLI and publishing tools
        wrapper._seren_agent = {
            "name": name.strip(),
            "description": description,
            "price": price,
            "compute_backend": compute_backend,
            "function_name": func.__name__,
        }

        return wrapper

    return decorator


def get_agent_metadata(func: Callable) -> Optional[dict]:
    """Extract Seren agent metadata from a decorated function.

    Args:
        func: A function potentially decorated with @agent

    Returns:
        Agent metadata dict or None if not a Seren agent
    """
    return getattr(func, "_seren_agent", None)


def is_seren_agent(func: Callable) -> bool:
    """Check if a function is a Seren agent template.

    Args:
        func: Function to check

    Returns:
        True if the function is decorated with @agent
    """
    return hasattr(func, "_seren_agent")

File: javascript/src/agent.js
/**
 * Agent decorator and metadata utilities.
 *
 * @module seren-agent/agent
 */

/** Symbol for agent metadata (prevents accidental access) */
const SEREN_AGENT_SYMBOL = Symbol.for("seren.agent");

/**
 * Create a Seren agent template.
 *
 * Wraps an async function with metadata for the Seren Store.
 * The agent is backend-agnostic - same code runs on any compute provider.
 *
 * @param {Object} config - Agent configuration
 * @param {string} config.name - Display name for the agent in the store
 * @param {string} [config.description] - Description of what the agent does
 * @param {string} config.price - Price per invocation in USD (e.g., "0.05")
 * @param {string} [config.computeBackend] - Preferred compute backend (e.g., "daytona", "modal")
 * @param {function} fn - The agent function implementing your logic
 * @returns {function} Wrapped agent function with metadata
 *
 * @example
 * ```javascript
 * import { agent } from 'seren-agent';
 *
 * export const run = agent({
 *   name: "Web Researcher",
 *   description: "Research topics via web search",
 *   price: "0.05"
 * }, async (input) => {
 *   const { query } = input;
 *   // Your agent logic here
 *   return { summary: "...", sources: [...] };
 * });
 * ```
 */
export function agent(config, fn) {
    const { name, description = "", price, computeBackend } = config;

    if (!name) {
        throw new Error("Agent name is required");
    }
    if (!price) {
        throw new Error("Agent price is required");
    }
    if (typeof fn !== "function") {
        throw new Error("Agent function is required as second argument");
    }

    // Create wrapper that preserves the original function
    async function wrappedAgent(input) {
        return fn(input);
    }

    // Attach metadata
    const metadata = {
        name,
        description,
        price,
        computeBackend,
    };

    wrappedAgent._seren_agent = metadata;
    wrappedAgent[SEREN_AGENT_SYMBOL] = metadata;

    return wrappedAgent;
}

/**
 * Get metadata from a Seren agent function.
 *
 * @param {function} fn - Agent function to inspect
 * @returns {Object|undefined} Agent metadata or undefined if not a Seren agent
 *
 * @example
 * ```javascript
 * import { getAgentMetadata } from 'seren-agent';
 *
 * const metadata = getAgentMetadata(run);
 * console.log(metadata.name);  // "Web Researcher"
 * console.log(metadata.price); // "0.05"
 * ```
 */
export function getAgentMetadata(fn) {
    if (typeof fn !== "function") {
        return undefined;
    }
    return fn[SEREN_AGENT_SYMBOL] || fn._seren_agent;
}

/**
 * Check if a function is a Seren agent.
 *
 * @param {function} fn - Function to check
 * @returns {boolean} True if the function is a Seren agent
 *
 * @example
 * ```javascript
 * import { isSerenAgent } from 'seren-agent';
 *
 * if (isSerenAgent(run)) {
 *   console.log("This is a Seren agent!");
 * }
 * ```
 */
export function isSerenAgent(fn) {
    return getAgentMetadata(fn) !== undefined;
}
